python/lsst/daf/butler/_quantum_backed.py

```diff
@@ -749,19 +745,40 @@ def _to_uuid_set(uuids: Iterable[str | uuid.UUID]) -> set[uuid.UUID]:
             """
             return {uuid.UUID(id) if isinstance(id, str) else id for id in uuids}
 
-        data = QuantumProvenanceData.__new__(cls)
-        setter = object.__setattr__
-        setter(data, "predicted_inputs", _to_uuid_set(predicted_inputs))
-        setter(data, "available_inputs", _to_uuid_set(available_inputs))
-        setter(data, "actual_inputs", _to_uuid_set(actual_inputs))
-        setter(data, "predicted_outputs", _to_uuid_set(predicted_outputs))
-        setter(data, "actual_outputs", _to_uuid_set(actual_outputs))
-        setter(
-            data,
-            "datastore_records",
-            {
-                key: SerializedDatastoreRecordData.direct(**records)
-                for key, records in datastore_records.items()
-            },
-        )
+        if PYDANTIC_V2:
+            data = cls.model_construct(
+                _fields_set={
+                    "predicted_inputs",
+                    "available_inputs",
+                    "actual_inputs",
+                    "predicted_outputs",
+                    "actual_outputs",
+                    "datastore_records",
+                },
+                predicted_inputs=_to_uuid_set(predicted_inputs),
+                available_inputs=_to_uuid_set(available_inputs),
+                actual_inputs=_to_uuid_set(actual_inputs),
+                predicted_outputs=_to_uuid_set(predicted_outputs),
+                actual_outputs=_to_uuid_set(actual_outputs),
+                datastore_records={
+                    key: SerializedDatastoreRecordData.direct(**records)
+                    for key, records in datastore_records.items()
+                },
+            )
+        else:
+            data = QuantumProvenanceData.__new__(cls)
+            setter = object.__setattr__
+            setter(data, "predicted_inputs", _to_uuid_set(predicted_inputs))
+            setter(data, "available_inputs", _to_uuid_set(available_inputs))
+            setter(data, "actual_inputs", _to_uuid_set(actual_inputs))
+            setter(data, "predicted_outputs", _to_uuid_set(predicted_outputs))
+            setter(data, "actual_outputs", _to_uuid_set(actual_outputs))
+            setter(
+                data,
+                "datastore_records",
+                {
+                    key: SerializedDatastoreRecordData.direct(**records)
+                    for key, records in datastore_records.items()
+                },
+            )
```

Comment:
Would it be possible to implement V1 `model_construct` to do the same thing as we do here to avoid if/else in every model class?

Reply:
There is v1 `BaseModel.construct()` which does the same thing as `model_construct`. We need to understand why @natelust decided not to use it. If we know why and need to change we would need to go back to the explicit setters above and change to using the `__pydantic_fields_set__` property instead.